mikutter: 4.0.0 -> 4.1.4

midchildan opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
* Closes #114440
* Fix broken build
* Add `passthru.updateScript`

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

midchildan committed:
mikutter: 4.0.0 -> 4.1.4

Closes #114440

r-rmcgibbo commented:
Result of `nixpkgs-review pr 119454` at dcc47e76 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built successfully:</summary>
  <ul>
    <li>mikutter</li>
    <li>rubyPackages.gtk2 (rubyPackages_2_6.gtk2)</li>
    <li>rubyPackages.pango (rubyPackages_2_6.pango)</li>
    <li>rubyPackages_2_7.gtk2</li>
    <li>rubyPackages_2_7.pango</li>
    <li>rubyPackages_3_0.gtk2</li>
    <li>rubyPackages_3_0.pango</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

which is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/networking/instant-messengers/mikutter/default.nix:54:3:
```
   |
54 |   buildInputs = [
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

desktop-file-utils is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/networking/instant-messengers/mikutter/default.nix:54:3:
```
   |
54 |   buildInputs = [
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/applications/networking/instant-messengers/mikutter/default.nix:101:3:
```
    |
101 |   meta = with lib; {
    |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/applications/networking/instant-messengers/mikutter/default.nix:67:3:
```
   |
67 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

midchildan commented:
Applied suggestions above.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/default.nix:
```diff
@@ -1,82 +1,109 @@
+  version = "4.1.4";
 
+  gems = bundlerEnv {
+    name = "mikutter-${version}-gems";
```
```suggestion
    name = "mikutter-gems";
```

Otherwise we cannot reuse the gems if the version changes.

midchildan replied:
Fixed.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/default.nix:
```diff
@@ -1,82 +1,109 @@
-# rm mikutter.4.0.0.tar.gz
-# rm gemset.nix Gemfile.lock; nix-shell -p bundler bundix --run 'bundle lock && bundix'
+let
+  version = "4.1.4";
 
```
```suggestion
```

midchildan replied:
Fixed.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/default.nix:
```diff
@@ -1,82 +1,109 @@
 stdenv.mkDerivation rec {
   pname = "mikutter";
-  version = "4.0.0";
+  inherit version;
```
```suggestion
  version = "4.1.4";
```

midchildan replied:
Fixed.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/default.nix:
```diff
@@ -1,82 +1,109 @@
-    unpackFile $src
-    rm -rf vendor
-  '';
+  postUnpack = "rm -rf vendor";
```
```suggestion
  postUnpack = ''
    rm -r vendor
  '';
```

midchildan replied:
Fixed.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/default.nix:
```diff
@@ -1,82 +1,109 @@
+      --set-key Exec --set-value $out/bin/mikutter \
+      --set-key Icon --set-value $out/opt/mikutter/core/skin/data/icon.png \
+      deployment/appimage/mikutter.desktop
+    rm -rv README LICENSE deployment
```
```suggestion
    rm -r README LICENSE deployment
```

midchildan replied:
Fixed.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/default.nix:
```diff
@@ -1,82 +1,109 @@
     install -v -D -m644 README $out/share/doc/mikutter/README
     install -v -D -m644 LICENSE $out/share/doc/mikutter/LICENSE
-    rm -v README LICENSE
+    desktop-file-install --dir $out/share/applications \
+      --set-key Exec --set-value $out/bin/mikutter \
+      --set-key Icon --set-value $out/opt/mikutter/core/skin/data/icon.png \
+      deployment/appimage/mikutter.desktop
```
Can we use makeDesktopItems here instead?

midchildan replied:
Switched to `makeDesktopItem`.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/default.nix:
```diff
@@ -1,82 +1,109 @@
+      deployment/appimage/mikutter.desktop
+    rm -rv README LICENSE deployment
+
+    cp -rv . $out/opt/mikutter
```
```suggestion
    cp -r . $out/opt/mikutter
```

midchildan replied:
Fixed.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/default.nix:
```diff
@@ -1,82 +1,109 @@
   '';
 
+  passthru.updateScript = import ./update.nix {
+    inherit lib pname version writers bundix bundler common-updater-scripts curl
+      jq;
```
```suggestion
    inherit lib pname version writers bundix bundler common-updater-scripts curl jq;
```

midchildan replied:
Fixed.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/update.nix:
```diff
@@ -0,0 +1,56 @@
+{ lib
+, pname
+, version
+, writers
+, bundix
+, bundler
+, common-updater-scripts
+, curl
+, jq
+}:
+
+let
+  binPath = lib.makeBinPath [ bundler bundix curl jq common-updater-scripts ];
+in
+writers.writeBash "update-mikutter" ''
```
Why are you not using a script wit nix-shell?

midchildan replied:
Mostly because it's easy to embed Nix variables. I've now changed it to a nix-shell script instead.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/update.nix:
```diff
@@ -0,0 +1,56 @@
+    # updateScript.
+    SRC="$(nix-build ../../../../.. --no-out-link -A mikutter.src)"
+
+    if [[ -d deps ]]; then
+      rm -rf deps
+    fi
```
```suggestion
    rm -rf deps
```

midchildan replied:
Fixed.

SuperSandro2000 commented on pkgs/development/ruby-modules/gem-config/default.nix:
```diff
@@ -25,7 +25,8 @@
 , file, libvirt, glib, vips, taglib, libopus, linux-pam, libidn, protobuf, fribidi, harfbuzz
 , bison, flex, pango, python3, patchelf, binutils, freetds, wrapGAppsHook, atk
-, bundler, libsass, libselinux ? null, libsepol ? null, shared-mime-info
+, bundler, libsass, libselinux ? null, libsepol ? null, shared-mime-info, libthai
```
```suggestion
, bundler, libsass, libselinux, libsepol, shared-mime-info, libthai
```

midchildan replied:
Fixed.

tomberek commented:
Result of `nixpkgs-review pr 119454` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>mikutter</li>
    <li>rubyPackages.gtk2 (rubyPackages_2_6.gtk2)</li>
    <li>rubyPackages.pango (rubyPackages_2_6.pango)</li>
    <li>rubyPackages_2_7.gtk2</li>
    <li>rubyPackages_2_7.pango</li>
    <li>rubyPackages_3_0.gtk2</li>
    <li>rubyPackages_3_0.pango</li>
  </ul>
</details>

midchildan committed:
fixup! mikutter: 4.0.0 -> 4.1.4

midchildan committed:
fixup! fixup! mikutter: 4.0.0 -> 4.1.4

midchildan committed:
fixup! fixup! fixup! mikutter: 4.0.0 -> 4.1.4

midchildan commented:
In addition to addressing review comments, I've made the following additional changes:

- Fixed Darwin build
- Added an `.app` bundle for Darwin
- Added an `installCheck` test

The test probably wouldn't catch errors originating from the most fragile dependencies, namely ruby-gnome gems, but it's better than nothing.

SuperSandro2000 commented on pkgs/development/ruby-modules/gem-config/default.nix:
```diff
@@ -453,6 +461,11 @@ in
       xorg.libpthreadstubs
       xorg.libXdmcp
+    ] ++ lib.optionals stdenv.isDarwin [ darwin.DarwinTools ];
+    buildInputs = [ libdatrie libthai ] ++ lib.optionals stdenv.isLinux [
```
```suggestion
    buildInputs = [ libdatrie libthai ]
      ++ lib.optionals stdenv.isLinux [
```

midchildan replied:
Fixed.

SuperSandro2000 commented on pkgs/development/ruby-modules/gem-config/default.nix:
```diff
@@ -25,7 +25,7 @@
 , file, libvirt, glib, vips, taglib, libopus, linux-pam, libidn, protobuf, fribidi, harfbuzz
 , bison, flex, pango, python3, patchelf, binutils, freetds, wrapGAppsHook, atk
-, bundler, libsass, libselinux ? null, libsepol ? null, shared-mime-info
+, bundler, libsass, libselinux, libsepol, shared-mime-info, libthai, libdatrie
```
```suggestion
, bundler, libsass, libselinux, libsepol, shared-mime-info, libthai, libdatrie, DarwinTools
```
Please do not use darwin directly as an input.

midchildan replied:
Fixed other gems as well.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/update.sh:
```diff
@@ -0,0 +1,51 @@
+    fi
+
+    latestVer="$(queryLatestVersion)"
+    if [[ "$currentVer" == "$latestVer" ]]; then
```
```suggestion
    if [[ $currentVer == $latestVer ]]; then
```

midchildan replied:
Dropping quotes would result in incorrect behavior in some cases because expanded variables on the right hand side of the comparison would be interpreted as a glob pattern when using the `[[` builtin. For example:

```bash
currentVer='1.0.0'
latestVer='1*'
if [[ $currentVer == $latestVer ]]; then
  echo 'up to date'
fi
```

The above script would output "up to date" because the `*` in latestVer is interpreted as a wildcard character.

Though quotes should be unnecessary for the left hand side, I still think it's a good idea to just quote it anyways because missing quotes do more harm than redundant ones when dealing with shell scripts.

SuperSandro2000 replied:
I thought I already have shellcheck remembered but I don't.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mikutter/update.sh:
```diff
@@ -0,0 +1,51 @@
+#!nix-shell -i bash -p bundler bundix curl jq common-updater-scripts
+# shellcheck shell=bash
+
+set -euxo pipefail
```
```suggestion
set -euo pipefail
```

midchildan replied:
Fixed.

midchildan committed:
fixup! fixup! fixup! fixup! mikutter: 4.0.0 -> 4.1.4

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 119454` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>mikutter</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
mikutter:
<details>

warning: EvalError
Cannot evaluate attribute ‘mikutter’ in ‘/nix/store/s5sl2b8l2sg9psn9s38xjqa0shailj9c-nixpkgs’.
</details>
</details>

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 119454` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
7 packages built:
</summary>
<ul>
<li>mikutter</li>
<li>rubyPackages.gtk2 (rubyPackages_2_6.gtk2)</li>
<li>rubyPackages.pango (rubyPackages_2_6.pango)</li>
<li>rubyPackages_2_7.gtk2</li>
<li>rubyPackages_2_7.pango</li>
<li>rubyPackages_3_0.gtk2</li>
<li>rubyPackages_3_0.pango</li>
</ul>
</details>
